libraries/chain/include/eosio/chain/incremental_merkle.hpp

```diff
@@ -6,212 +6,70 @@
-         bool partial = false;
-         auto max_depth = detail::calculate_max_depth(_node_count + 1);
-         auto current_depth = max_depth - 1;
-         auto index = _node_count;
-         auto top = digest;
-         auto active_iter = _active_nodes.begin();
-         auto updated_active_nodes = vector<DigestType>();
-         updated_active_nodes.reserve(max_depth);
-
-         while (current_depth > 0) {
-            if (!(index & 0x1)) {
-               // we are collapsing from a "left" value and an implied "right" creating a partial node
-
-               // we only need to append this node if it is fully-realized and by definition
-               // if we have encountered a partial node during collapse this cannot be
-               // fully-realized
-               if (!partial) {
-                  updated_active_nodes.emplace_back(top);
-               }
-
-               // calculate the partially realized node value by implying the "right" value is identical
-               // to the "left" value
-               if constexpr (canonical) {
-                  top = DigestType::hash(make_canonical_pair(top, top));
-               } else {
-                  top = DigestType::hash(std::make_pair(std::cref(top), std::cref(top)));
-               }
-               partial = true;
-            } else {
-               // we are collapsing from a "right" value and an fully-realized "left"
-
-               // pull a "left" value from the previous active nodes
-               const auto& left_value = *active_iter;
-               ++active_iter;
-
-               // if the "right" value is a partial node we will need to copy the "left" as future appends still need it
-               // otherwise, it can be dropped from the set of active nodes as we are collapsing a fully-realized node
-               if (partial) {
-                  updated_active_nodes.emplace_back(left_value);
-               }
-
-               // calculate the node
-               if constexpr (canonical) {
-                  top = DigestType::hash(make_canonical_pair(left_value, top));
-               } else {
-                  top = DigestType::hash(std::make_pair(std::cref(left_value), std::cref(top)));
-               }
-            }
-
-            // move up a level in the tree
-            --current_depth;
-            index = index >> 1;
-         }
-
-         // append the top of the collapsed tree (aka the root of the merkle)
-         updated_active_nodes.emplace_back(top);
-
-         // store the new active_nodes
-         detail::move_nodes(_active_nodes, std::move(updated_active_nodes));
-
-         // update the node count
-         ++_node_count;
-
-         return _active_nodes.back();
-
-      }
-
-      /**
-       * return the current root of the incremental merkle
-       */
-      DigestType get_root() const {
-         if (_node_count > 0) {
-            return _active_nodes.back();
+class incremental_merkle_tree {
```

Comment:
A class comment would be nice.

Reply:
done, please let me know if you would add/change something.